props/saber_fett263_buttons.h

```diff
@@ -293,39 +659,745 @@ class SaberFett263Buttons : public virtual PropBase {
+        break;
+      case CLASH_LOCKUP:
+        if (battle_mode_) clash_value = (SaberBase::GetClashStrength() - saved_gesture_control.clashdetect) / saved_gesture_control.maxclash;
+        if (SFX_bgnlock) {
+          files = SFX_bgnlock.files_found() - 1;
+          file_number = clamp(files * clash_value, 0, files);
+        }
+        SaberBase::SetLockup(SaberBase::LOCKUP_NORMAL);
+        SaberBase::DoBeginLockup();
+        break;
+      case CLASH_LOCKUP_END:
+        if (SFX_endlock) {
+          files = SFX_endlock.files_found() - 1;
+          file_number = clamp(files * (fusor.swing_speed() / 600), 0, files);
+        }
+        SaberBase::DoEndLockup();
+        SaberBase::SetLockup(SaberBase::LOCKUP_NONE);
+        break;
+    }
+    clash_type_ = CLASH_NONE;
+  }
+  #endif
+
   void Loop() override {
     PropBase::Loop();
     DetectTwist();
     Vec3 mss = fusor.mss();
+    if (clash_type_ != CLASH_NONE && millis() - clash_impact_millis_ > 1) {
+      // CHECK PUSH
+      if (clash_type_ == CLASH_CHECK) {
+        Event(BUTTON_NONE, EVENT_PUSH);
+        clash_type_ = CLASH_NONE;
+      }
+      if (clash_type_ != CLASH_LOCKUP_END) {
+#ifdef FETT263_CLASH_STRENGTH_SOUND
+        SelectClashFile();
+#else
+      if (SaberBase::GetClashStrength() < saved_gesture_control.clashdetect) {
+        SaberBase::DoClash();
+        clash_type_ = CLASH_NONE;
+      } else {
+        SaberBase::SetLockup(SaberBase::LOCKUP_NORMAL);
+        SaberBase::DoBeginLockup();
+        auto_lockup_on_ = true;
+        clash_type_ = CLASH_NONE;
+      }
+#endif
+      }
+    }
+    sound_queue_.Poll(wav_player);
+#ifdef FETT263_EDIT_MODE_MENU      
+    if (next_event_ && !wav_player->isPlaying()) {
+      if (rehearse_) {
+        if (SaberBase::IsOn()) {
+          rehearse_ = false;
+          Off();
+#ifdef FETT263_SAVE_CHOREOGRAPHY
+          SaveChoreo();
+#endif
+        } else {
+          PlayMenuSound("rehrsbgn.wav");
+          FastOn();
+        }
+        next_event_ = false;
+      } else {
+        switch (menu_type_) {
+          case MENU_IGNITION_TIME:
+          case MENU_RETRACTION_TIME:
+            next_event_ = false;
+            SetInOut();
+            break;
+          default:
+            next_event_ = false;
+            break;
+        }
+      }
+    }
+    if (off_event_ && millis() - last_rotate_ > 200) {
+      Off();
+      off_event_ = false;
+      restart_ = true;
+      last_rotate_ = millis();
+    }
+    if (restart_ && millis() - last_rotate_ > calc_ + 1000) {
+      restart_ = false;
+      FastOn();
+    }
+#endif
+#ifdef FETT263_EDIT_MODE_MENU
+      if (edit_color_ && color_mode_ != COLOR_LIST) {
+#define H_CHANGE (M_PI / 98304)
+#define L_ANGLE (M_PI / 30)
+#define H_ANGLE (M_PI / 16384)
+#define EDIT_MODE_ZOOM (M_PI * 2 / 2000)
+      float a = fusor.angle2() - hsl_angle_;
+      if (a > M_PI) a-=M_PI*2;      
+      if (a < -M_PI) a+=M_PI*2;
+      switch (color_mode_) {
```

Comment:
I think the next step to simplifying this is to observe that all of these things share a lot of code, but the angle is different.

So we can do this:

float angle = 100;
switch (color_mode_) {
   case EDIT_COLOR: angle = H_ANGLE; break;
   case ZOOM_COLOR: angle = EDIT_MODE_ZOOM; break;
  case EDIT_WHITE: 
   case EDIT_BLACK: angle = L_ANGLE; break;
   default: break;
}

int steps = 0;
 if (a > angle * 2/3) {
            hsl_angle_ += angle;
            if (hsl_angle_ > M_PI) hsl_angle_ -= M_PI * 2;
            steps++;
          }
          if (a < -angle * 2/3) {
            hsl_angle_ -= angle;
            if (hsl_angle_ < M_PI) hsl_angle_ += M_PI * 2;
            steps--;
          }

switch (color_mode_) {
   case EDIT_COLOR:
   case ZOOM_COLOR:
        hsl_.H = fract(hsl.H + H_CHANGE * steps);
        break;
    case EDIT_WHITE:
        if (step < 0) {.... }
        if (step > 0) { .... }
        break;
    case EDIT_BLACK:
        if (step < 0) { ... }
        if (step > 0) { .... }
        break;
}

Then we'll see if maybe we can simplify even more after that. :)

Reply:
thanks, updated working as expected.